pkg/inject/inject.go

```diff
@@ -384,15 +384,56 @@ func (conf *ResourceConfig) CreateDefaultOpaquePortsPatch() ([]byte, error) {
 			return nil, err
 		}
 	} else if !conf.HasWorkloadAnnotation(k8s.ProxyOpaquePortsAnnotation) {
-		opaquePorts := conf.GetValues().Proxy.OpaquePorts
-		patch, err = conf.CreateAnnotationPatch(opaquePorts)
+		defaultPorts := strings.Split(conf.GetValues().Proxy.OpaquePorts, ",")
+		var filteredPorts []string
+		// If the workload is a pod, then only add the default opaque ports
+		// that it exposes as container ports.
+		//
+		// If the workload is a service, then only add the default opaque
+		// ports that are exposed as a service port, or targeted as a
+		// targetPort.
+		if conf.IsPod() {
+			filteredPorts = conf.FilterPodOpaquePorts(defaultPorts)
+		} else if conf.IsService() {
+			service := conf.workload.obj.(*corev1.Service)
+			for _, p := range service.Spec.Ports {
+				port := strconv.Itoa(int(p.Port))
+				if util.ContainsString(port, defaultPorts) {
+					filteredPorts = append(filteredPorts, port)
+				} else if util.ContainsString(strconv.Itoa(int(p.TargetPort.IntVal)), defaultPorts) {
```

Comment:
worth remembering that we miss the case where we target a named port and a pod has a container port with that name which is a default opaque port.  But I don't think there's really anything we can do about that, users will need to manually annotate their service in that case.

Reply:
Yep I think we should be able to address this as part of #6783. When that check runs we'll be able to match all pods and services, and should be able to issue a warning that a service is missing a port in its annotation because it's `targetPort` is a name.